compiler/src/dotty/tools/dotc/transform/Dependencies.scala

```diff
@@ -184,23 +185,29 @@ abstract class Dependencies(root: ast.tpd.Tree, @constructorOnly rootContext: Co
     def setLogicOwner(local: Symbol) =
       val encClass = local.owner.enclosingClass
       val preferEncClass =
-        (
           encClass.isStatic
             // non-static classes can capture owners, so should be avoided
           && (encClass.isProperlyContainedIn(local.topLevelClass)
               // can be false for symbols which are defined in some weird combination of supercalls.
               || encClass.is(ModuleClass, butNot = Package)
                   // needed to not cause deadlocks in classloader. see t5375.scala
               )
-        )
-        || (
+          && (!sym.isAnonymousFunction || sym.owner.ownersIterator.exists(_.isConstructor))
+            // For anonymous functions in static objects, we prefer them to be static because
```

Comment:
What does the "in static objects" part mean here?

Reply:
Static means toplevel or nested in a static object.

Reply:
How does the comment relate to an anonymous function in the extended test case? It is in a class, not a static object. I feel like I'm missing something here...

```scala
class C {
  def x(): Int => String = (i: Int) => i.toString
}
```

Reply:
It added more comments. Hope it is clearer now.

Reply:
Thanks, that helps me understand.